File: src/tools/pullImage.ts
import type { Tool, ToolProgress } from "src/Tool";
import { createImageValidator } from "src/services/docker/imageValidator";
import { z } from "zod";

export const PullImageInputSchema = z.object({ image: z.string() });
export type PullImageInput = z.infer<typeof PullImageInputSchema>;

export interface PullImageResult {
  ok: boolean;
  status: "valid" | "invalid" | "unknown";
  source?: "local" | "registry" | "pulled" | "unavailable";
  error?: string;
  suggestion?: string;
}

export const pullImage: Tool<PullImageInput, PullImageResult> = {
  name: "pull_image",
  description:
    "Validate a Docker image reference and pre-pull it when it exists in a registry but is not local.",
  inputSchema: PullImageInputSchema,
  category: "escape-hatch",
  needsPermission: () => true,
  call: async function* (input, ctx): AsyncGenerator<ToolProgress, PullImageResult> {
    const validator = ctx.imageValidator ?? createImageValidator(ctx.dockerEngine);
    yield { type: "progress", msg: `Validating ${input.image}...` };
    const validation = await validator.validateImage(input.image, { signal: ctx.abortSignal });

    if (validation.status === "invalid") {
      return {
        ok: false,
        status: "invalid",
        source: validation.source,
        ...(validation.error ? { error: validation.error } : {}),
        ...(validation.suggestion ? { suggestion: validation.suggestion } : {}),
      };
    }

    if (validation.status === "unknown") {
      return {
        ok: true,
        status: "unknown",
        source: validation.source,
        ...(validation.error ? { error: validation.error } : {}),
      };
    }

    if (validation.source === "registry") {
      if (!ctx.dockerEngine.pullImage) {
        return {
          ok: false,
          status: "valid",
          source: "registry",
          error: "Docker engine does not support image pulling",
        };
      }
      yield { type: "progress", msg: `Pulling ${input.image}...` };
      for await (const line of ctx.dockerEngine.pullImage(input.image, {
        signal: ctx.abortSignal,
      })) {
        yield { type: "progress", msg: line };
      }
      return { ok: true, status: "valid", source: "pulled" };
    }

    return { ok: true, status: "valid", source: validation.source };
  },
};

File: src/tools/pull_image.py
"""pull_image tool.

Parity: ``src/tools/pullImage.ts``.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Literal

from pydantic import BaseModel

from src.services.docker.image_validator import create_image_validator
from src.tool import ToolContext, ToolDone, ToolProgress


class PullImageInput(BaseModel):
    image: str


class PullImageResult(BaseModel):
    ok: bool
    status: Literal["valid", "invalid", "unknown"]
    source: Literal["local", "registry", "pulled", "unavailable"] | None = None
    error: str | None = None
    suggestion: str | None = None


class _PullImageTool:
    name = "pull_image"
    description = (
        "Validate a Docker image reference and pre-pull it when it exists in a registry "
        "but is not local."
    )
    input_schema = PullImageInput
    category = "escape-hatch"

    def needs_permission(self, _input: PullImageInput) -> bool:
        return True

    async def call(
        self, input: PullImageInput, ctx: ToolContext
    ) -> AsyncIterator[ToolProgress | ToolDone]:
        validator = (
            ctx.image_validator
            if ctx.image_validator is not None
            else create_image_validator(ctx.docker_engine)
        )
        yield ToolProgress(msg=f"Validating {input.image}...")
        validation = await validator.validate_image(
            input.image, signal=ctx.abort_signal
        )

        if validation.status == "invalid":
            yield ToolDone(
                PullImageResult(
                    ok=False,
                    status="invalid",
                    source=validation.source,
                    error=validation.error,
                    suggestion=validation.suggestion,
                )
            )
            return

        if validation.status == "unknown":
            yield ToolDone(
                PullImageResult(
                    ok=True,
                    status="unknown",
                    source=validation.source,
                    error=validation.error,
                )
            )
            return

        if validation.source == "registry":
            pull_image_fn = getattr(ctx.docker_engine, "pull_image", None)
            if pull_image_fn is None:
                yield ToolDone(
                    PullImageResult(
                        ok=False,
                        status="valid",
                        source="registry",
                        error="Docker engine does not support image pulling",
                    )
                )
                return
            yield ToolProgress(msg=f"Pulling {input.image}...")
            async for line in pull_image_fn(input.image, signal=ctx.abort_signal):
                yield ToolProgress(msg=line)
            yield ToolDone(PullImageResult(ok=True, status="valid", source="pulled"))
            return

        yield ToolDone(
            PullImageResult(
                ok=True,
                status="valid",
                source=validation.source,
            )
        )


pull_image = _PullImageTool()

__all__ = ["PullImageInput", "PullImageResult", "pull_image"]